### ISSUE TITLE
ArgumentException thrown when multiple tasks observe a metric in parallel

### ISSUE THREAD

**Issue by FlorianHockmann** [PrometheusClientNet/Prometheus.Client#132]
We recently updated some of our micro services from Prometheus.Client 3.1.0 to 4.2.0 and from .NET Core 3.1 to .NET 5. This resulted in frequent exceptions being thrown by Prometheus.Client.

Two example stack traces:

```c#
 System.ArgumentException: Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.
    at System.Array.Sort[T](T[] array, Int32 index, Int32 length, IComparer`1 comparer)
    at Prometheus.Client.SummaryImpl.QuantileStream.FlushBuffer(Span`1 destination, Int32& bufferSize)
    at Prometheus.Client.SummaryImpl.QuantileStream.FlushBuffer()
    at Prometheus.Client.Summary.ForkState(Int64& count, Double& sum, Double[] values)
    at Prometheus.Client.Summary.Collect(IMetricsWriter writer)
    at Prometheus.Client.MetricFamily`4.Prometheus.Client.Collectors.Abstractions.ICollector.Collect(IMetricsWriter writer)
    at Prometheus.Client.Collectors.CollectorRegistry.CollectToAsync(IMetricsWriter writer)
    at Prometheus.Client.ScrapeHandler.ProcessAsync(ICollectorRegistry registry, Stream outputStream)
    at Prometheus.Client.AspNetCore.PrometheusExtensions.<>c__DisplayClass1_0.<<UsePrometheusServer>b__3>d.MoveNext()
--- End of stack trace from previous location ---
    at Microsoft.AspNetCore.Builder.Extensions.MapMiddleware.Invoke(HttpContext context)
    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

```c#
System.ArgumentException: Offset and length were out of bounds for the array or count is greater than the number of elements from index to the end of the source collection.
    at System.Array.Sort[T](T[] array, Int32 index, Int32 length, IComparer`1 comparer)
    at Prometheus.Client.SummaryImpl.QuantileStream.FlushBuffer(Span`1 destination, Int32& bufferSize)
    at Prometheus.Client.SummaryImpl.QuantileStream.FlushBuffer()
    at Prometheus.Client.SummaryImpl.QuantileStream.Append(Double value)
    at Prometheus.Client.Summary.Observe(Double val)
    at Motor.Extensions.Hosting.Publisher.TypedMessagePublisher`2.PublishMessageAsync(MotorCloudEvent`1 cloudEvent, CancellationToken token)
    at Motor.Extensions.Hosting.MultiOutputServiceAdapter`2.HandleMessageAsync(MotorCloudEvent`1 dataCloudEvent, CancellationToken token)
    at Motor.Extensions.Hosting.Internal.PrepareDelegatingMessageHandler`1.HandleMessageAsync(MotorCloudEvent`1 dataCloudEvent, CancellationToken token)
```

The exception however only occurs if we use multiple tasks in parallel that can observe metrics. So, I suspect that this is a problem with concurrency.

**Comment by sanych-sun** [comment]
Thank you for reporting this. Will take a look ASAP.

**Comment by sanych-sun** [comment]
I've created PR to get this sorted. Also added some basic parallel tests to avoid such cases in future. The change will be included into upcoming 4.3.0 release.

**Comment by phnx47** [comment]
Released in v4.3.0

**Comment by FlorianHockmann** [comment]
Thanks a lot for fixing this so quickly!
